<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>pafr: reading, manipulating and plotting genome alignments in the PAF format • pafr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="pafr: reading, manipulating and plotting genome alignments in the PAF format">
<meta property="og:description" content="pafr">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pafr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/dwinter/pafr/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><link href="Introduction_to_pafr_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet">
<script src="Introduction_to_pafr_files/anchor-sections-1.0/anchor-sections.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>pafr: reading, manipulating and plotting genome alignments in the PAF format</h1>
                        <h4 class="author">David Winter</h4>
            
            <h4 class="date">2020-12-02</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/dwinter/pafr/blob/master/vignettes/Introduction_to_pafr.Rmd"><code>vignettes/Introduction_to_pafr.Rmd</code></a></small>
      <div class="hidden name"><code>Introduction_to_pafr.Rmd</code></div>

    </div>

    
    
<div id="reading-in-data-with-read_paf" class="section level2">
<h2 class="hasAnchor">
<a href="#reading-in-data-with-read_paf" class="anchor"></a>Reading in data with <code>read_paf</code>
</h2>
<p><code>pafr</code> processes data stored in the <strong>P</strong>airwise m<strong>A</strong>pping <strong>F</strong>ormat (PAF), as produced by <code>minimap2</code> and other whole-genome aligners. PAF is a plain text tabular format where each row represents an alignment between two sequences. Each <code>.paf</code> file has at least the following 12 columns.</p>
<table class="table">
<thead><tr class="header">
<th><strong>Column</strong></th>
<th><strong>Name</strong></th>
<th><strong>Data Type</strong></th>
<th><strong>Description</strong></th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>qname</td>
<td>string</td>
<td>Query sequence name</td>
</tr>
<tr class="even">
<td>2</td>
<td>qlen</td>
<td>int</td>
<td>Query sequence length</td>
</tr>
<tr class="odd">
<td>3</td>
<td>qstart</td>
<td>int</td>
<td>Query start coordinate (0-based)</td>
</tr>
<tr class="even">
<td>4</td>
<td>qend</td>
<td>int</td>
<td>Query end coordinate (0-based)</td>
</tr>
<tr class="odd">
<td>5</td>
<td>strand</td>
<td>char</td>
<td>‘+’ if query/target on the same strand; ‘-’ if opposite</td>
</tr>
<tr class="even">
<td>6</td>
<td>tname</td>
<td>string</td>
<td>Target sequence name</td>
</tr>
<tr class="odd">
<td>7</td>
<td>tlen</td>
<td>int</td>
<td>Target sequence length</td>
</tr>
<tr class="even">
<td>8</td>
<td>tstart</td>
<td>int</td>
<td>Target start coordinate on the original strand</td>
</tr>
<tr class="odd">
<td>9</td>
<td>tend</td>
<td>int</td>
<td>Target end coordinate on the original strand</td>
</tr>
<tr class="even">
<td>10</td>
<td>nmatch</td>
<td>int</td>
<td>Number of matching bases in the mapping</td>
</tr>
<tr class="odd">
<td>11</td>
<td>alen</td>
<td>int</td>
<td>Number of bases, including gaps, in the mapping</td>
</tr>
<tr class="even">
<td>12</td>
<td>mapq</td>
<td>int</td>
<td>Mapping quality (0-255, with 255 if missing)</td>
</tr>
</tbody>
</table>
<p>In addition, each row can have a variable number of 'tags', which share a similar format to those used in SAM files <a href="https://samtools.github.io/hts-specs/SAMtags.pdf">See the SAM specification (pdf)</a>.</p>
<p><code>pafr</code> provides the function <code>read_paf</code> to read <code>.paf</code> alignments into an R session.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dwinter.github.io/pafr/">pafr</a></span><span class="op">)</span></code></pre></div>
<pre><code>## Loading required package: ggplot2</code></pre>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">path_to_fungal_alignment</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"fungi.paf"</span>, package <span class="op">=</span> <span class="st">"pafr"</span><span class="op">)</span>
<span class="va">ali</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_paf.html">read_paf</a></span><span class="op">(</span><span class="va">path_to_fungal_alignment</span><span class="op">)</span></code></pre></div>
<p>By default, the object returned by <code>read_paf</code> behaves almost exactly like a base R <code>data.frame</code>. In fact, the only difference is that this object prints a nice summary of the object contents when called directly or printed (rather than the thousands of lines the file might contain):</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ali</span></code></pre></div>
<pre><code>## pafr object with 2501 alignments (36.5Mb)
##  8 query seqs
##  8 target seqs
##  11 tags: NM, ms, AS, nn, tp, cm, s1, s2, dv, cg, zd</code></pre>
<p>For the worked example, a total of 35 Mb of aligned genome sequence is represented by 8 query and 8 target sequences. The 12 standard parts of the PAF file are represented by columns named as in the table above.</p>
<p>Any tags present in the <code>.paf</code> file are included as an additional column. The tags used by <code>minimap2</code> <a href="https://lh3.github.io/minimap2/minimap2.html">are documented here</a> and also in the <code>man</code> page for that program.</p>
<p>One interesting tag for this particular alignment is <code>dv</code>, which encodes the approximate per-base difference ('divergence') between the query and target sequences. We can use this tag to visualize the relationship between alignment length and sequence divergence. Because the alignment information is stored in a <code>data.frame</code>, we can pass this information directly to <code>ggplot2</code>:</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://rpkgs.datanovia.com/ggpubr/">ggpubr</a></span><span class="op">)</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">ali</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">alen</span>, <span class="va">dv</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.6</span>, colour<span class="op">=</span><span class="st">"steelblue"</span>, size<span class="op">=</span><span class="fl">2</span><span class="op">)</span> <span class="op">+</span> 
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span><span class="op">(</span><span class="st">"Alignment length (kb)"</span>, label <span class="op">=</span>  <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">/</span> <span class="fl">1e3</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span><span class="st">"Per base divergence"</span><span class="op">)</span> <span class="op">+</span> 
    <span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/theme_pubr.html">theme_pubr</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning: Removed 52 rows containing missing values (geom_point).</code></pre>
<p><img src="Introduction_to_pafr_files/figure-html/alen-1.png" width="576" style="display: block; margin: auto;"></p>
<p>We can see that there are many short alignments, some of which are very divergent. But there are also some very long alignments, all of which show high similarity.</p>
<p>Because the <code>pafr</code> object is effectively a <code>data.frame</code>, we can again use standard R functions to inspect or analyse it. For example, we can calculate the mean divergence level for alignments featuring each sequence in the query genome.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">by_q</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/aggregate.html">aggregate</a></span><span class="op">(</span><span class="va">dv</span> <span class="op">~</span> <span class="va">qname</span>, data<span class="op">=</span><span class="va">ali</span>, FUN<span class="op">=</span><span class="va">mean</span><span class="op">)</span>
<span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="va">by_q</span><span class="op">)</span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">qname</th>
<th align="right">dv</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Q_chr1</td>
<td align="right">0.0514848</td>
</tr>
<tr class="even">
<td align="left">Q_chr2</td>
<td align="right">0.0493231</td>
</tr>
<tr class="odd">
<td align="left">Q_chr3</td>
<td align="right">0.0556407</td>
</tr>
<tr class="even">
<td align="left">Q_chr4</td>
<td align="right">0.0543924</td>
</tr>
<tr class="odd">
<td align="left">Q_chr5</td>
<td align="right">0.0501364</td>
</tr>
<tr class="even">
<td align="left">Q_chr6</td>
<td align="right">0.0560656</td>
</tr>
<tr class="odd">
<td align="left">Q_chr7</td>
<td align="right">0.0479018</td>
</tr>
<tr class="even">
<td align="left">Q_chrm</td>
<td align="right">0.0185438</td>
</tr>
</tbody>
</table>
<p>Interestingly enough, <code>Q_chrm</code> is the mitochondrial genome, so it appears that the mitochondrial genome displays less divergence than any of the autosomal chromosomes in this particular species.</p>
<div id="reading-really-big-alignments" class="section level3">
<h3 class="hasAnchor">
<a href="#reading-really-big-alignments" class="anchor"></a>Reading really big alignments</h3>
<p>Though <code>read_paf</code> has been optimised to process the tags in a paf file quickly, it can still take some time to read paf files that have a large number of alignments. For example, it takes about 30 seconds to read an alignment between two fragmented vertebrate genomes on my desktop PC. If speed is of the essence and you know you don't need the data encoded in the tags, there are a few tricks to get a <code>pafr</code> object more quickly.</p>
<p>First, you can tell <code>read_paf</code> to ignore the tag information by setting the <code>include_tags</code> argument to <code>FALSE</code>. Let's use the <code>microbenchmark</code> benchmark package to to see how much time we can save by skipping the tags in our example alignment.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">microbenchmark</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/microbenchmark/man/microbenchmark.html">microbenchmark</a></span><span class="op">(</span>
  tags <span class="op">=</span> <span class="fu"><a href="../reference/read_paf.html">read_paf</a></span><span class="op">(</span><span class="va">path_to_fungal_alignment</span><span class="op">)</span>,
  no_tags <span class="op">=</span> <span class="fu"><a href="../reference/read_paf.html">read_paf</a></span><span class="op">(</span><span class="va">path_to_fungal_alignment</span>, include_tags<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>,
  times<span class="op">=</span><span class="fl">10</span>
<span class="op">)</span></code></pre></div>
<pre><code>## Unit: milliseconds
##     expr       min        lq      mean    median        uq       max neval
##     tags 196.55810 197.10971 199.71339 199.00400 201.69211 205.82663    10
##  no_tags  41.33416  41.88992  44.34422  43.27138  47.22024  48.20145    10</code></pre>
<p>That's about five times faster on my computer. Huge or highly-fragmented genomes may still produce so many alignments that they take a long time to read even with this approach. If you run into this issue, you may want to remove the tag data from the file before processing it in R. For example, you could use the unix utility <code>cut</code> to take only the first 12 columns from a file.</p>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash"><span class="fu">cut</span> -f1-12 [alignment_with_tags.paf] <span class="op">&gt;</span> tagless.paf</code></pre></div>
<p>With not tag data to consider, <code>read_paf</code> will be able to deal with this data more quickly. <code>pafr</code> also provides the function <code>as_paf</code> to allow users to convert a 12-column <code>data.frame</code> or similar into a <code>pafr</code> object that can be used by visualisation functions in this package. This approach allows you to use <code>read.table</code> or optimised functions like <code><a href="https://readr.tidyverse.org/reference/read_delim.html">readr::read_delim</a></code> or <code><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html">data.table::fread</a></code> to process large files.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tagless_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="st">"tagless.paf"</span><span class="op">)</span>
<span class="va">tagless_ali</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as_paf.html">as_paf</a></span><span class="op">(</span><span class="va">tagless_df</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="filtering-and-subsetting-alignments" class="section level2">
<h2 class="hasAnchor">
<a href="#filtering-and-subsetting-alignments" class="anchor"></a>Filtering and subsetting alignments</h2>
<p>Often the first thing you will want to do after reading in data is to drop low-quality, short or otherwise non-ideal alignments. <code>minimap2</code> uses the <code>tp</code> ('type of alignment') tag to mark secondary alignments (<em>i.e.</em>, possible alignments between the query and target sequences that are not the best possible alignments for those regions). The function <code>filter_secondary_alignments</code> removes non-primary alignments from a <code>pafr</code> object.</p>
<p>We can use this function to filter our <code>ali</code> object and see how many alignments (<em>i.e.</em>, rows of data) are removed as a result:</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">prim_alignment</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_secondary_alignments.html">filter_secondary_alignments</a></span><span class="op">(</span><span class="va">ali</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">ali</span><span class="op">)</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">prim_alignment</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 339</code></pre>
<p>In some cases, you may want to remove alignments with low mapping quality, high divergence, short length or some other specific property. Again, because the <code>pafr</code> object behaves like a <code>data.frame</code>, you can use your favourite base <code>R</code> or <code>tidyverse</code> functions to do this.</p>
<p>Here, we are removing short alignments and those with low mapping quality scores:</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">long_ali</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">ali</span>, <span class="va">alen</span> <span class="op">&gt;</span> <span class="fl">1e4</span> <span class="op">&amp;</span> <span class="va">mapq</span> <span class="op">&gt;</span> <span class="fl">40</span><span class="op">)</span>
<span class="va">long_ali</span></code></pre></div>
<pre><code>## pafr object with 419 alignments (26.8Mb)
##  8 query seqs
##  8 target seqs
##  11 tags: NM, ms, AS, nn, tp, cm, s1, s2, dv, cg, zd</code></pre>
<p>Note that this operation has removed quite a large amount of data (thousands of alignments and ~10 Mb of sequence). It is important to keep track of these kinds of data losses after filtering.</p>
</div>
<div id="plotting-alignments" class="section level2">
<h2 class="hasAnchor">
<a href="#plotting-alignments" class="anchor"></a>Plotting alignments</h2>
<p>A major aim of <code>pafr</code> is to generate compelling visualisations of genomic alignments. To this end, the package implements three types of genomic graphics, making use of the <code>ggplot2</code> plotting environment.</p>
<p>Although <code>ggplot2</code> is a powerful visualisation tool, we found that producing high-quality plots of genome alignments was difficult, requiring the use of many different layers and multiple transformations to datasets. The visualisation functions provided by <code>pafr</code> automate the process of transforming data into a useable form and adding layers to plots. Each of these functions returns <code>ggplot2</code> plots, which can then be modified by the addition of further ggplot layers, scales and themes.</p>
<div id="dot-plot" class="section level3">
<h3 class="hasAnchor">
<a href="#dot-plot" class="anchor"></a>Dot plot</h3>
<p>Often the most useful visualization of a genome alignment is the <a href="https://en.wikipedia.org/wiki/Dot_plot_(bioinformatics)">dot plot</a>. <code>pafr</code> makes it easy to produce a dot plot from a <code>.paf</code> alignment:</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/dotplot.html">dotplot</a></span><span class="op">(</span><span class="va">prim_alignment</span><span class="op">)</span></code></pre></div>
<p><img src="Introduction_to_pafr_files/figure-html/dotplot_plain-1.png" width="576" style="display: block; margin: auto;"></p>
<div id="altering-the-order-of-sequences-in-the-dot-plot" class="section level4">
<h4 class="hasAnchor">
<a href="#altering-the-order-of-sequences-in-the-dot-plot" class="anchor"></a>Altering the order of sequences in the dot plot</h4>
<p>The default plot is quite sparse, with each aligned segment shown as a dark line and the borders of the sequences in the query and target genomes as dashed lines. Additional arguments to <code>dotplot</code> let us modify the plot.</p>
<p>For instance, we can add labels for each query and target sequence (<code>label_seqs</code>) and changes the order in which target sequences appear. Because the dot plot produced here is a ggplot object, we can use <code><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw()</a></code> to change the plot theme too.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/dotplot.html">dotplot</a></span><span class="op">(</span><span class="va">prim_alignment</span>, label_seqs<span class="op">=</span><span class="cn">TRUE</span>, order_by<span class="op">=</span><span class="st">"qstart"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="Introduction_to_pafr_files/figure-html/dot_qstart-1.png" width="576" style="display: block; margin: auto;"></p>
<p>The argument <code>order_by</code> takes three possible values: 'size', 'qstart' or 'provided'. 'Size' (the default value) simply lines up query and target sequences from largest to smallest. 'qstart' keeps the query sequences ordered by size, but rearranges the targets by where they match to query sequences. For example, <code>T_chr5</code> is moved to the second sequence in the plot above, as it matches the first query chromosome. If <code>ordered_by</code> is set to 'provided', you need to provide a list to the function with two elements: the order of the query, and then target, sequences. This approach can also be used to down-sample an alignment to only a few focal sequences.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">to_keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>
        <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Q_chr1"</span>, <span class="st">"Q_chr5"</span>, <span class="st">"Q_chr4"</span>, <span class="st">"Q_chr6"</span><span class="op">)</span>,
        <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"T_chr2"</span>, <span class="st">"T_chr5"</span>, <span class="st">"T_chr3"</span>, <span class="st">"T_chr6"</span><span class="op">)</span>
<span class="op">)</span>
<span class="fu"><a href="../reference/dotplot.html">dotplot</a></span><span class="op">(</span><span class="va">prim_alignment</span>, label_seqs<span class="op">=</span><span class="cn">TRUE</span>, order_by<span class="op">=</span><span class="st">"provided"</span>, ordering<span class="op">=</span><span class="va">to_keep</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning: Dropping data from sequences absent from ordering:
##  Q_chr2,Q_chr3,Q_chr7,Q_chrm</code></pre>
<pre><code>## Warning: Dropping data from sequences absent from ordering:
##  T_chr1,T_chr7,T_chr4,T_chrm</code></pre>
<p><img src="Introduction_to_pafr_files/figure-html/dot_provided-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div id="adding-annotations" class="section level4">
<h4 class="hasAnchor">
<a href="#adding-annotations" class="anchor"></a>Adding annotations</h4>
<p>You may have additional information about one or both of your genomes that you want annotate on your dot plots.</p>
<p>For instance, here we know the locations of the centromeres in the query genome. We can use <code>read_bed</code> to load these locations into <code>pafr</code>.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">path_to_centro</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Q_centro.bed"</span>, package <span class="op">=</span> <span class="st">"pafr"</span><span class="op">)</span>
<span class="va">centro</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_bed.html">read_bed</a></span><span class="op">(</span><span class="va">path_to_centro</span><span class="op">)</span>
<span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">centro</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">chrom</th>
<th align="right">start</th>
<th align="right">end</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Q_chr1</td>
<td align="right">1305119</td>
<td align="right">1325119</td>
</tr>
<tr class="even">
<td align="left">Q_chr2</td>
<td align="right">3311000</td>
<td align="right">3331000</td>
</tr>
<tr class="odd">
<td align="left">Q_chr5</td>
<td align="right">1780000</td>
<td align="right">1800000</td>
</tr>
<tr class="even">
<td align="left">Q_chr6</td>
<td align="right">1005337</td>
<td align="right">1025337</td>
</tr>
<tr class="odd">
<td align="left">Q_chr4</td>
<td align="right">951000</td>
<td align="right">971000</td>
</tr>
<tr class="even">
<td align="left">Q_chr7</td>
<td align="right">449000</td>
<td align="right">469000</td>
</tr>
</tbody>
</table>
<p>In <code>ggplot</code> style, we use the <code><a href="https://rdrr.io/r/base/Arithmetic.html">+</a></code> symbol to add the results of <code>highlight_query</code> to a dot plot. (The ordering of the chromosomes in the plot is inherited from the dotplot, so does not need to be specified in the highlight function).</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/dotplot.html">dotplot</a></span><span class="op">(</span><span class="va">prim_alignment</span>, <span class="st">"qstart"</span><span class="op">)</span> <span class="op">+</span> 
    <span class="fu"><a href="../reference/highlight_query.html">highlight_query</a></span><span class="op">(</span>bed<span class="op">=</span><span class="va">centro</span><span class="op">)</span></code></pre></div>
<p><img src="Introduction_to_pafr_files/figure-html/highlight-1.png" width="576" style="display: block; margin: auto;"></p>
<p>It is not apparent from the very small intervals plotted above, but 'highlights' here are filled rectangles. Plotting a larger rectangle for the target genome makes this clearer.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">interval</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>chrom<span class="op">=</span><span class="st">"T_chr3"</span>, start<span class="op">=</span><span class="fl">2000000</span>, end<span class="op">=</span><span class="fl">3000000</span><span class="op">)</span>
<span class="fu"><a href="../reference/dotplot.html">dotplot</a></span><span class="op">(</span><span class="va">prim_alignment</span>, label_seqs<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> 
    <span class="fu"><a href="../reference/highlight_query.html">highlight_target</a></span><span class="op">(</span><span class="va">interval</span><span class="op">)</span></code></pre></div>
<p><img src="Introduction_to_pafr_files/figure-html/highlight_t-1.png" width="576" style="display: block; margin: auto;"></p>
<p>The 'highlight' functions use ggplot's <code>geom_rect</code> to produce these annotations, and will pass on any argument use by this function (e.g. colour, fill, alpha...).</p>
</div>
</div>
<div id="synteny-plot" class="section level3">
<h3 class="hasAnchor">
<a href="#synteny-plot" class="anchor"></a>Synteny plot</h3>
<p>The dot plot gives us a 'whole genome' view of a genomic alignment. Very often, however, we will also want to zoom in to look at how specific regions of two particular chromosomes are aligned to each other. The function <code>plot_synteny</code> provides a way to visualize alignments at this level.</p>
<p>For example, we can compare regions of query chromosome 3 and target chromosome 4.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_synteny.html">plot_synteny</a></span><span class="op">(</span><span class="va">long_ali</span>, q_chrom<span class="op">=</span><span class="st">"Q_chr3"</span>, t_chrom<span class="op">=</span><span class="st">"T_chr4"</span>, centre<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="Introduction_to_pafr_files/figure-html/unnamed-chunk-2-1.png" width="576" style="display: block; margin: auto;"></p>
<p>In this plot, each sequence is represented by a white box and each alignment is a grey segment connecting those sequences that align to each other. In this case, the alignment is quite straightforward. However, the plot looks less clear when most of the alignments are in reverse-and-complement format.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_synteny.html">plot_synteny</a></span><span class="op">(</span><span class="va">long_ali</span>, q_chrom<span class="op">=</span><span class="st">"Q_chr5"</span>, t_chrom<span class="op">=</span><span class="st">"T_chr5"</span>, centre<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="Introduction_to_pafr_files/figure-html/unnamed-chunk-3-1.png" width="576" style="display: block; margin: auto;"></p>
<p>The simple solution in these cases is to set the argument <code>rc</code> to <code>TRUE</code>, forcing the target sequence to be flipped.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_synteny.html">plot_synteny</a></span><span class="op">(</span><span class="va">long_ali</span>, q_chrom<span class="op">=</span><span class="st">"Q_chr5"</span>, t_chrom<span class="op">=</span><span class="st">"T_chr5"</span>, centre<span class="op">=</span><span class="cn">TRUE</span>, rc<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="Introduction_to_pafr_files/figure-html/unnamed-chunk-4-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div id="coverage-plot" class="section level3">
<h3 class="hasAnchor">
<a href="#coverage-plot" class="anchor"></a>Coverage plot</h3>
<p>The function <code>plot_coverage</code> gives a useful way to see how much of a given genome is included in a genome alignment. By default, it displays each sequence in the target chromosome as a rectangular box, with shaded regions representing parts of the target genome that are included in an alignment:</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_coverage.html">plot_coverage</a></span><span class="op">(</span><span class="va">prim_alignment</span><span class="op">)</span>   </code></pre></div>
<p><img src="Introduction_to_pafr_files/figure-html/cov_plot-1.png" width="576" style="display: block; margin: auto;"></p>
<p>It is also possible to produce the same plot using the query genome as the reference.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_coverage.html">plot_coverage</a></span><span class="op">(</span><span class="va">prim_alignment</span>, target<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>   </code></pre></div>
<p><img src="Introduction_to_pafr_files/figure-html/cov_target-1.png" width="576" style="display: block; margin: auto;"></p>
<div id="coverage-by-query-sequence" class="section level4">
<h4 class="hasAnchor">
<a href="#coverage-by-query-sequence" class="anchor"></a>Coverage by query sequence</h4>
<p>Sometimes, in addition to knowing how much of a given genome is represented in a whole genome alignment, it can be interesting to know which sequences from the other genome are aligning. For example, we might want to 'paint' regions of the target sequence by which query sequence they form an alignment to. The <code>plot_coverage</code> function can shade the alignment blocks it produces according to any column in the alignment.</p>
<p>For example, this is how we could paint genomes for query sequence name (<em>i.e.</em>, the <code>qname</code> column), taking advantage of <code>ggplot</code>'s <code>scale_colour_brewer</code> to use a nice colour palette.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_coverage.html">plot_coverage</a></span><span class="op">(</span><span class="va">prim_alignment</span>, fill<span class="op">=</span><span class="st">'qname'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_fill_brewer</a></span><span class="op">(</span>palette<span class="op">=</span><span class="st">"Set1"</span><span class="op">)</span></code></pre></div>
<p><img src="Introduction_to_pafr_files/figure-html/cov_colourful-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
</div>
</div>
<div id="feedback-bugs-and-development" class="section level1">
<h1 class="hasAnchor">
<a href="#feedback-bugs-and-development" class="anchor"></a>Feedback, bugs and development</h1>
<p>We would love to hear your feedback on <code>pafr</code>. You can file issues (including bugs, usage questions and feature requests) at the package's github issue-tracker: <a href="https://github.com/dwinter/pafr/issues" class="uri">https://github.com/dwinter/pafr/issues</a></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by David Winter.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
